pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go

```diff
@@ -87,38 +114,43 @@ type kubenetNetworkPlugin struct {
 	hostportSyncer  hostport.HostportSyncer
 	hostportManager hostport.HostPortManager
 	iptables        utiliptables.Interface
+	iptablesv6      utiliptables.Interface
 	sysctl          utilsysctl.Interface
 	ebtables        utilebtables.Interface
 	// binDirs is passed by kubelet cni-bin-dir parameter.
 	// kubenet will search for CNI binaries in DefaultCNIDir first, then continue to binDirs.
-	binDirs           []string
-	nonMasqueradeCIDR string
-	podCidr           string
-	gateway           net.IP
+	binDirs            []string
+	nonMasqueradeCIDRs []string
+	podCIDRs           []*net.IPNet
+	podGateways        []net.IP
 }
 
 func NewPlugin(networkPluginDirs []string) network.NetworkPlugin {
-	protocol := utiliptables.ProtocolIpv4
 	execer := utilexec.New()
 	dbus := utildbus.New()
-	sysctl := utilsysctl.New()
-	iptInterface := utiliptables.New(execer, dbus, protocol)
+	iptInterface := utiliptables.New(execer, dbus, utiliptables.ProtocolIpv4)
+	iptInterfacev6 := utiliptables.New(execer, dbus, utiliptables.ProtocolIpv6)
 	return &kubenetNetworkPlugin{
-		podIPs:            make(map[kubecontainer.ContainerID]string),
-		execer:            utilexec.New(),
-		iptables:          iptInterface,
-		sysctl:            sysctl,
-		binDirs:           append([]string{DefaultCNIDir}, networkPluginDirs...),
-		hostportSyncer:    hostport.NewHostportSyncer(iptInterface),
-		hostportManager:   hostport.NewHostportManager(iptInterface),
-		nonMasqueradeCIDR: "10.0.0.0/8",
+		podIPs:          make(map[kubecontainer.ContainerID]utilsets.String),
+		execer:          utilexec.New(),
+		iptables:        iptInterface,
+		iptablesv6:      iptInterfacev6,
+		sysctl:          utilsysctl.New(),
+		binDirs:         append([]string{DefaultCNIDir}, networkPluginDirs...),
+		hostportSyncer:  hostport.NewHostportSyncer(iptInterface),
+		hostportManager: hostport.NewHostportManager(iptInterface),
+		//TODO (khenidak): maybe it should be defaulted to zero address?
+		// 10.0.0.0/8 is the default value before dualstack changes
+		nonMasqueradeCIDRs: []string{"10.0.0.0/8"},
+		podCIDRs:           make([]*net.IPNet, 0),
+		podGateways:        make([]net.IP, 0),
 	}
 }
 
-func (plugin *kubenetNetworkPlugin) Init(host network.Host, hairpinMode kubeletconfig.HairpinMode, nonMasqueradeCIDR string, mtu int) error {
+func (plugin *kubenetNetworkPlugin) Init(host network.Host, hairpinMode kubeletconfig.HairpinMode, nonMasqueradeCIDRs string, mtu int) error {
 	plugin.host = host
 	plugin.hairpinMode = hairpinMode
-	plugin.nonMasqueradeCIDR = nonMasqueradeCIDR
+	plugin.nonMasqueradeCIDRs = strings.Split(nonMasqueradeCIDRs, ",")
```

Comment:
Question for the followers, especially folks with direct v6 experience:  Should we even bother with this?  My thinking:

1) kubenet is more or less deprectaed
2) the masquerade flag is single, so is under-powered for real v4 usage
3) IMO making kubelet configure masquerade was a mistake

The v4 masquerade was required for providers that only egress VM IPs (e.g. GCP) and is mostly used for that or for users who do clever stuff with IP reuse between clusters.  For those latter users, this is not sufficiently powerful, so we send them to the ip-masq-agent as a daemonset.  Even for cloud providers it is not sufficient - it doesn't even cover RFC-1918.

So if most people don't use this flag anyway, should we bother to fix it.  We could instead say "if you need masquerade, you should set it up outside of kubernetes proper - run an agent, configure init scripts, etc."

Thoughts?

@dcbw and @squeed ?

Reply:
I think you're right, for a few reasons:
- Anybody who is doing anything this complicated is almost certainly not using kubenet
- Kubenet can't be long for this world anyways
- Most importantly, masquerading for ipv6 is very silly indeed.

So yes, this flag should be marked deprecated, and I see no need for it to have v6 support extended.

Reply:
The original thought was to go alpha (no production workloads) on Kubenet, as outlined in the KEP. This will not require any CNI (outside k/k monorepo) to do any work, which would have required serious coordination across two repos. This is probably the only reason why kubenet is still needed. 

As for masquerading, in absolute yes it is silly. But then again kubenet works on fictitious address range that the world outside the cluster does not know about, hence the need for masquerading and accordingly non masquerading. Btw, it is optional not mandatory. i.e. it is turned off by default, users can use `ip-masq-agent` (or CNI own) to do the needed routing/masquerading

Reply:
On one hand, I'd like to kill the masquerade support and simplify this all.

On the other hand, you'll still have to add support for dual-stack iptables support, so it's not simplifying much.

I'll leave it up to you.  I would accept not updating this flag (except to deprecate or comment v4 only) or I will accept implementing it for a single v6.

Reply:
I'm fine with either one, but slight preference to just not updating it.